modules/ws/src/main/java/org/mule/module/ws/consumer/MuleWSDLLocator.java

```diff
@@ -0,0 +1,163 @@
+ * where wsdl4j has to retrieve a resource (imported XSD, WSDL)
+ */
+public class MuleWSDLLocator implements WSDLLocator
+{
+    private static final Logger logger = LoggerFactory.getLogger(WSConsumer.class);
+
+    private String baseURI;
+    private String latestImportedURI;
+    private boolean useConnectorToRetrieveWsdl;
+    private HttpRequesterWsdlRetrieverStrategy httpRetrieverStrategy;
+    private URLRetrieverStrategy urlRetrieverStrategy;
+    private Collection<InputStream> streams = new ArrayList<InputStream>();
+    
+    public MuleWSDLLocator(MuleWSDLLocatorConfig config) throws MuleException
+    {
+        this.baseURI = config.getBaseURI();
+        this.useConnectorToRetrieveWsdl = config.isUseConnectorToRetrieveWsdl();
+        createURLRetrieverStrategy();
+        if (useConnectorToRetrieveWsdl)
+        {
+            createHttpRequesterWsdlRetrieverStrategy(config.getTlsContextFactory(), config.getProxyConfig(), config.getContext());
+        }
+    }
+
+
+    private void createURLRetrieverStrategy()
+    {
+        urlRetrieverStrategy = new URLRetrieverStrategy();
+
+    }
+
+
+    private void createHttpRequesterWsdlRetrieverStrategy(TlsContextFactory tlsContextFactory, ProxyConfig proxyConfig, MuleContext context) throws MuleException
+    {
+        httpRetrieverStrategy = new HttpRequesterWsdlRetrieverStrategy(tlsContextFactory, proxyConfig, context);
+        httpRetrieverStrategy.startHttpClient();
+    }
+
+
+    @Override
+    public InputSource getBaseInputSource()
+    {
+        try
+        {
+            return getInputSource(baseURI);
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public InputSource getImportInputSource(String parentLocation, String importLocation)
+    {
+
+        try
+        {
+            latestImportedURI = IOUtils.getResourceAsUrl(importLocation, getClass()).toString();
+
+            return getInputSource(importLocation);
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    private InputSource getInputSource(String url) throws WSDLException
+    {
+        boolean isHttpRequester = verifyIsHttpRequester(url);
+
+        InputStream resultStream = null;
+
+        if (useConnectorToRetrieveWsdl && isHttpRequester)
+        {
+            resultStream = httpRetrieverStrategy.retrieveWsdlResource(url);
+        }
+        else
+        {
+            resultStream = urlRetrieverStrategy.retrieveWsdlResource(url);
+        }
+
+        streams.add(resultStream);
+
+        return new InputSource(resultStream);
+    }
+
+    @Override
+    public String getBaseURI()
+    {
+        return baseURI;
+    }
+
+    @Override
+    public String getLatestImportURI()
+    {
+        return latestImportedURI;
+    }
+
+    @Override
+    public void close()
+    {
+        if (useConnectorToRetrieveWsdl)
```

Comment:
I move the clean logic to the strategy. As I create the strategy each time the WSDLLocator method to create a InputSource is invoked, the httpClient is not reused